File: scala/bob/src/main/scala/Bob.scala
object Bob {
  private val SMALL_LETTERS: String = "qwertzuiopasdfghjklyxcvbnm";
  private val LETTERS: String = "qwertzuiopasdfghjklyxcvbnmQWERTZUIOPASDFGHJKLYXCVBNM"

  def response(statement: String): String = {
    if (isSilence(statement)){
      return "Fine. Be that way!"
    }
    if (isShoutingQuestion(statement)){
      return "Calm down, I know what I'm doing!"
    }
    if (isShouting(statement)){
      return "Whoa, chill out!"
    }
    if (isQuestion(statement)){
      return "Sure."
    }
    return "Whatever."
  }

  def isShouting(statement: String): Boolean = {
    var wasLetter: Boolean = false;
    for (letter <- statement){
      if (SMALL_LETTERS.contains(letter)){
        return false
      }
      if (LETTERS.contains(letter)){
        wasLetter = true
      }
    }
    return wasLetter
  }

  def isQuestion(statement: String): Boolean = {
    return statement.trim().endsWith("?")
  }

  def isShoutingQuestion(statement: String): Boolean = {
    return isQuestion(statement) && isShouting(statement)
  }

  def isSilence(statement: String): Boolean = {
    return statement.trim().isEmpty()
  }
}

File: java/bob/src/main/java/Bob.java
class Bob {

    String hey(String input) {
        if (input.trim().isEmpty()) {
            return "Fine. Be that way!";
        }

        boolean isYelling = input.chars().allMatch(c -> !Character.isLetter(c) || Character.isUpperCase(c)) && input.chars().anyMatch(Character::isLetter);
        boolean isQuestion = input.trim().endsWith("?");

        if (isYelling && isQuestion) {
            return "Calm down, I know what I'm doing!";
        }

        if (isQuestion) {
            return "Sure.";
        }

        if (isYelling) {
            return "Whoa, chill out!";
        }

        return "Whatever.";
    }

}